src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -61,19 +79,37 @@ public static class ActivatorUtilities
                     {
                         bestLength = length;
                         bestMatcher = matcher;
+                        multipleBestLengthFound = false;
+                    }
+                    else if (bestLength == length)
+                    {
+                        multipleBestLengthFound = true;
                     }
 
                     seenPreferred |= isPreferred;
                 }
+
+                if (bestLength != -1)
+                {
+                    if (multipleBestLengthFound)
+                    {
+                        throw new InvalidOperationException(SR.Format(SR.MultipleCtorsFoundWithBestLength, instanceType, bestLength));
+                    }
+
+                    return bestMatcher.CreateInstance(provider);
+                }
             }
 
-            if (bestLength == -1)
+            var argumentTypes = new Type[parameters.Length];
+            for (int i = 0; i < argumentTypes.Length; i++)
             {
-                string? message = $"A suitable constructor for type '{instanceType}' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.";
-                throw new InvalidOperationException(message);
+                argumentTypes[i] = parameters[i].GetType();
             }
 
-            return bestMatcher.CreateInstance(provider);
+            FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructorInfo, out int?[] parameterMap);
```

Comment:
Do we have unit tests that cover the cases mentioned previously? Specifically the cases where `IServiceProviderIsService` returns `false` incorrectly.

Reply:
added in https://github.com/dotnet/runtime/pull/75846/commits/6e46bd11ed915b24118055eb5a62d10754c00b18